spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/health/PubSubHealthIndicator.java

```diff
@@ -16,55 +16,174 @@
 import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import org.springframework.util.concurrent.ListenableFuture;
 
 /**
- * Default implemenation of
+ * Default implementation of
  * {@link org.springframework.boot.actuate.health.HealthIndicator} for Pub/Sub. Validates
- * if connection is successful by looking for a random generated subscription name that
- * won't be found. If no subscription is found we know the client is able to connect to
- * GCP Pub/Sub APIs.
+ * if connection is successful by pulling messages from the pubSubTemplate using
+ * {@link PubSubTemplate#pullAsync(String, Integer, Boolean)}.
+ *
+ * <p>If a custom subscription has been specified, this health indicator will signal "up"
+ * if messages are successfully pulled and (optionally) acknowledged <b>or</b> if a
+ * successful pull is performed but no messages are returned from Pub/Sub.</p>
+ *
+ * <p>If no subscription has been specified, this health indicator will pull messages from a random subscription
+ * that is expected not to exist. It will signal "up" if it is able to connect to GCP Pub/Sub APIs,
+ * i.e. the pull results in a response of {@link StatusCode.Code#NOT_FOUND} or
+ * {@link StatusCode.Code#PERMISSION_DENIED}.</p>
+ *
+ * <p>Note that messages pulled from the subscription will not be acknowledged, unless you
+ * set the {@code acknowledgeMessages} option to "true". However, take care not to configure
+ * a subscription that has a business impact, or leave the custom subscription out completely.
  *
  * @author Vinicius Carvalho
+ * @author Patrik Hörlin
  *
  * @since 1.2.2
  */
 public class PubSubHealthIndicator extends AbstractHealthIndicator {
 
+	/**
+	 * Template used when performing health check calls.
+	 */
 	private final PubSubTemplate pubSubTemplate;
 
-	public PubSubHealthIndicator(PubSubTemplate pubSubTemplate) {
+	/**
+	 * Indicates whether a user subscription has been configured.
+	 */
+	private final boolean specifiedSubscription;
+
+	/**
+	 * Subscription used when health checking.
+	 */
+	private final String subscription;
+
+	/**
+	 * Timeout when performing health check.
+	 */
+	private final long timeoutMillis;
+
+	/**
+	 * If messages pull should be acknowledged or not.
+	 */
+	private final boolean acknowledgeMessages;
+
+	public PubSubHealthIndicator(PubSubTemplate pubSubTemplate, String healthCheckSubscription, long timeoutMillis, boolean acknowledgeMessages) {
 		super("Failed to connect to Pub/Sub APIs. Check your credentials and verify you have proper access to the service.");
-		Assert.notNull(pubSubTemplate, "PubSubTemplate can't be null");
+		Assert.notNull(pubSubTemplate, "pubSubTemplate can't be null");
 		this.pubSubTemplate = pubSubTemplate;
+		this.specifiedSubscription = StringUtils.hasText(healthCheckSubscription);
+		if (this.specifiedSubscription) {
+			this.subscription = healthCheckSubscription;
+		}
+		else {
+			this.subscription = UUID.randomUUID().toString();
```

Comment:
Subscription names must start with a letter. I'd suggest a prefix of "spring-cloud-gcp-healthcheck".
https://googleapis.dev/java/google-cloud-pubsub/latest/com/google/pubsub/v1/Subscription.Builder.html#setName-java.lang.String-